addons/point_of_sale/static/src/js/Popups/PosDialog.js

```diff
@@ -0,0 +1,43 @@
+odoo.define('point_of_sale.Dialog', function(require) {
+    'use strict';
+
+    const Registries = require('point_of_sale.Registries');
+    const PosComponent = require('point_of_sale.PosComponent');
+    const { useBus } = require("@web/core/utils/hooks");
+
+    class PosDialog extends PosComponent {
+        constructor() {
+            super(...arguments);
+            useBus(this.env.posbus, 'show-popup', this.__showPopup);
+            useBus(this.env.posbus, 'close-popup', this.__closePopup);
+            this.popups = [];
+        }
+        __showPopup(eventDetail) {
+            const { id, name, props, resolve } = eventDetail;
+            const component = this.constructor.components[name];
+            if (component.dontShow) {
+                resolve();
+                return;
+            }
+            this.popups.push({
+                name,
+                component,
+                props: Object.assign(props || {}, { id, resolve }),
+                key: `${name}-${id}`
+            });
+            this.render();
+        }
+        __closePopup(id) {
+            const index = this.popups.findIndex(popup => popup.props.id == id);
+            if (index != -1) {
+                this.popups.splice(index, 1);
+                this.render();
+            }
+        }
```

Comment:
That won't work if you're always expecting an `id`.
There are occurrences in the code where we trigger the `'close-popup'` event without calling the `cancel()` method, you'll need to refactor those occurrences to pass an id as well.
Or what you might want to use instead is a stack so that you don't need to keep track of id ?

Reply:
`cancel` and `confirm` of AwaitablePopup always provide the id. But that only means we need to always derive AwaitablePopup and call `cancel` or `confirm` if we want to define a new popup.

And yeah, I was actually thinking of stack for that matter, but I believe it's better if we know the identity of the popup that we clicked.

Reply:
You know what. This looks like an example where a unit test can really help because the component is relatively independent from the UI. But still, coupled with the `AwaitablePopup` and `PosComponent`.